File: web-core/skin/src/presets/client/convert.ts
import type { PresetKind, PresetKindState } from "./kinds.js";
import type { PresetHeader, PresetRecord } from "./record.js";
import type { WirePreset } from "./wire-preset.js";

export function text(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function toState<T>(kind: PresetKind, item: WirePreset): T {
  if (kind === "outfit") {
    return {
      palette: text(item.palette?.name),
      forms: (item.forms ?? []).map((form) => text(form.name)),
      tags: (item.tags ?? []).map((tag) => text(tag.name)),
      overrides: item.overrides,
      author: item.author,
    } as T;
  }

  if (kind === "palette") {
    return {
      scales: item.scales,
      dimensions: item.dimensions,
      light: item.light,
      dark: item.dark,
      author: item.author,
    } as T;
  }

  if (kind === "form") {
    return {
      component: item.component,
      recipe: item.recipe,
      keyframes: item.keyframes,
      variantTags: item.variantTags,
      author: item.author,
    } as T;
  }

  if (kind === "assembly") {
    return { component: item.component, assembly: item.assembly, author: item.author } as T;
  }

  if (kind === "tag") {
    return { label: item.tagLabel, author: item.author } as T;
  }

  return { component: item.component, data: item.data, author: item.author } as T;
}

export function toHeader(item: WirePreset): PresetHeader {
  const name = text(item.name);
  return {
    id: text(item.id),
    label: text(item.label) === "" ? name : text(item.label),
    name,
    kind: text(item.kind) as PresetKind,
    savedAt: text(item.savedAt),
  };
}

export function toRecord<K extends PresetKind>(kind: K, item: WirePreset): PresetRecord<PresetKindState[K]> {
  const header = toHeader(item);
  return { ...header, state: { name: header.name, ...toState<PresetKindState[K]>(kind, item) } };
}

File: backend/presets/internal/graphql/convert.go
package graphql

import (
	"encoding/json"
	"fmt"

	"presets/internal/graphql/model"
	"presets/internal/kinds"
	presetsmodel "presets/internal/model"
)

// toPreset строит типизированную GraphQL-модель ОДНОЙ записи через internal/kinds — тот же
// реестр, что валидирует запись на вход (bbolt-byte-invariant, ROADMAP.yaml). Связи
// (Outfit.palette/forms/tags) сюда намеренно не входят: они резолвятся отдельно, лениво, только
// когда клиент их спросил (см. outfit.resolvers.go).
func toPreset(record *presetsmodel.Record) (model.Preset, error) {
	decoded, ok := kinds.New(record.Kind)
	if !ok {
		return nil, fmt.Errorf("presets: запись %q несёт незарегистрированный вид %q", record.ID, record.Kind)
	}
	if err := json.Unmarshal(record.State, decoded); err != nil {
		return nil, fmt.Errorf("presets: запись %q (%s) не разобралась по своему виду: %w", record.ID, record.Kind, err)
	}

	switch state := decoded.(type) {
	case *kinds.Palette:
		return &model.Palette{
			Meta:       record.Meta,
			Author:     state.Author,
			Scales:     model.JSON(state.Scales),
			Dimensions: model.JSON(state.Dimensions),
			Light:      model.JSON(state.Light),
			Dark:       model.JSON(state.Dark),
		}, nil

	case *kinds.Form:
		return &model.Form{
			Meta:        record.Meta,
			Component:   state.Component,
			Recipe:      model.JSON(state.Recipe),
			Keyframes:   model.JSON(state.Keyframes),
			VariantTags: model.JSON(state.VariantTags),
			Author:      state.Author,
		}, nil

	case *kinds.Outfit:
		return &model.Outfit{
			Meta:        record.Meta,
			PaletteName: state.Palette,
			FormNames:   state.Forms,
			TagNames:    state.Tags,
			Overrides:   model.JSON(state.Overrides),
			Author:      state.Author,
		}, nil

	case *kinds.Content:
		return &model.Content{
			Meta:      record.Meta,
			Component: state.Component,
			Data:      model.JSON(state.Data),
			Author:    state.Author,
		}, nil

	case *kinds.Tag:
		var tagLabel *string
		if state.Label != "" {
			tagLabel = &state.Label
		}
		return &model.Tag{
			Meta:     record.Meta,
			TagLabel: tagLabel,
			Author:   state.Author,
		}, nil

	case *kinds.Api:
		return &model.Api{
			Meta:      record.Meta,
			Endpoints: model.JSON(state.Endpoints),
			Groups:    model.JSON(state.Groups),
			Defs:      model.JSON(state.Defs),
		}, nil

	case *kinds.Adapter:
		var extra *string
		if state.Extra != "" {
			extra = ptrString(state.Extra)
		}
		return &model.Adapter{
			Meta:      record.Meta,
			Root:      state.Root,
			Rules:     model.JSON(state.Rules),
			Extra:     extra,
			Providers: model.JSON(state.Providers),
			Consumers: model.JSON(state.Consumers),
		}, nil

	case *kinds.Menu:
		return &model.Menu{
			Meta:         record.Meta,
			AdapterNames: state.Adapters,
		}, nil

	case *kinds.ComponentAssembly:
		return &model.Assembly{
			Meta:      record.Meta,
			Component: state.Component,
			Assembly:  model.JSON(state.Assembly),
			Author:    state.Author,
		}, nil

	default:
		// Вид зарегистрирован в internal/kinds, но toPreset про него не знает — реестр и эта
		// функция разъехались (новый вид завели, а GraphQL-модель для него не подключили).
		return nil, fmt.Errorf("presets: вид %q зарегистрирован в kinds, но не подключён к GraphQL-модели", record.Kind)
	}
}
